### PR TITLE
Lower data-tiled multi_mma to intrinsics.

### PR DESCRIPTION
**Pull request by bjacob**
This lowers `multi_mma` with `DataTiledMMAAttr` down to `amdgpu` intrinsics.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 1]
```diff
@@ -903,10 +905,34 @@ std::tuple<int64_t, int64_t, int64_t> DataTiledMMAAttr::getMNKShape() const {
           opaqueLayout.kSize * getUnrollK()};
 }
 
+static SmallVector<int64_t>
+filterSwizzledShape(const TileSwizzle &swizzle,
+                    std::function<bool(TileSwizzle::Dim)> predicate) {
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: Add short doc.
Also, small nit: The term `filter` makes it seem like the result will be missing the "filtered" dims, but they are set to `1`. Maybe something like `sliceSwizzledShape`/`sliceSwizzledShapeDims`?

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 2]
```diff
@@ -903,10 +905,34 @@ std::tuple<int64_t, int64_t, int64_t> DataTiledMMAAttr::getMNKShape() const {
...
+    auto swizzle = getSwizzle(*this, fragment);
+    SmallVector<int64_t> shape =
+        filterSwizzledShape(swizzle, [](TileSwizzle::Dim d) {
+          return d.kind != TileSwizzle::Dim::Kind::CrossThread;
+        });
+    Type elementType = elementTypes[static_cast<int>(fragment)];
+    return VectorType::get(shape, elementType);
+  };
+  return {getVectorType(MMAFragment::Lhs), getVectorType(MMAFragment::Rhs),
+          getVectorType(MMAFragment::Acc)};
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: Maybe explicitly pass in `A`, `B`, and `C` element types? Then you don't need to static_cast the fragment to an index, and it only costs a few extra chars here to add the parameter.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 3]
```diff
@@ -917,6 +943,161 @@ FailureOr<IREE::GPU::MMAScope> DataTiledMMAAttr::getMmaScope() const {
   return IREE::GPU::MMAScope::Workgroup;
 }
 
+LogicalResult DataTiledMMAAttr::populateOperandOffsetsSizesStrides(
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp:943]: Nice! This function became very simple with the Dim::Kind enums.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 4]
```diff
@@ -917,6 +943,161 @@ FailureOr<IREE::GPU::MMAScope> DataTiledMMAAttr::getMmaScope() const {
...
+        loc, value, indices, internalShape, strides);
+    distributedValue = flattenVector(builder, loc, extract);
+    incrementIndices(indices, crossIntrinsicShape);
+  }
+  return distributedValues;
+}
+
+// Generates amdgpu.mfma/wmma operation on the given inputs for this attribute
+// type.
+FailureOr<Value> DataTiledMMAAttr::buildMmaOperation(OpBuilder &builder,
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: WMMA is not supported, so should probably fail in that case and indicate enabling it in the doc as TODO. Same with other functions here.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_tile_and_fuse.mlir [hunk 5]
```diff
@@ -718,3 +718,109 @@ hal.executable public @main {
...
+// CHECK-DAG:  %[[C_00_0:.+]] = vector.shape_cast %[[C_SLICE00]] : vector<1x1x1x1x1x4xf32> to vector<4xf32>
+// CHECK-DAG:  %[[C_01_0:.+]] = vector.shape_cast %[[C_SLICE01]] : vector<1x1x1x1x1x4xf32> to vector<4xf32>
+// CHECK-DAG:  %[[C_70_0:.+]] = vector.shape_cast %[[C_SLICE70]] : vector<1x1x1x1x1x4xf32> to vector<4xf32>
+// CHECK-DAG:  %[[C_71_0:.+]] = vector.shape_cast %[[C_SLICE71]] : vector<1x1x1x1x1x4xf32> to vector<4xf32>
+// CHECK-DAG:  %[[A_EXTRACT00:.+]] = vector.extract %[[A_READ]][0, 0, 0, 0] : f32 from vector<8x1x1x4xf32>
+// CHECK-DAG:  %[[A_EXTRACT01:.+]] = vector.extract %[[A_READ]][0, 0, 0, 1] : f32 from vector<8x1x1x4xf32>
+// CHECK-DAG:  %[[A_EXTRACT02:.+]] = vector.extract %[[A_READ]][0, 0, 0, 2] : f32 from vector<8x1x1x4xf32>
+// CHECK-DAG:  %[[A_EXTRACT03:.+]] = vector.extract %[[A_READ]][0, 0, 0, 3] : f32 from vector<8x1x1x4xf32>
+// CHECK-DAG:  %[[A_EXTRACT70:.+]] = vector.extract %[[A_READ]][7, 0, 0, 0] : f32 from vector<8x1x1x4xf32>
+// CHECK-DAG:  %[[A_EXTRACT71:.+]] = vector.extract %[[A_READ]][7, 0, 0, 1] : f32 from vector<8x1x1x4xf32>
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_tile_and_fuse.mlir:795]: You can add `CHECK-COUNT-24` between these to make sure there are the right number of slices without adding too many lines here, i.e.:
```
// CHECK-COUNT-24: vector.extract %[[A_READ]][{{[0-9]+}}, 0, 0, 1] : f32 from vector<8x1x1x4xf32>
```

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_tile_and_fuse.mlir [hunk 6]
```diff
@@ -718,3 +718,109 @@ hal.executable public @main {
...
+// CHECK-DAG:  %[[C_00_1:.+]] = amdgpu.mfma %[[A_EXTRACT00]] * %[[B_EXTRACT00]] + %[[C_00_0]]
+// CHECK-DAG:  %[[C_00_2:.+]] = amdgpu.mfma %[[A_EXTRACT01]] * %[[B_EXTRACT01]] + %[[C_00_1]]
+// CHECK-DAG:  %[[C_00_3:.+]] = amdgpu.mfma %[[A_EXTRACT02]] * %[[B_EXTRACT02]] + %[[C_00_2]]
+// CHECK-DAG:  %[[C_00_4:.+]] = amdgpu.mfma %[[A_EXTRACT03]] * %[[B_EXTRACT03]] + %[[C_00_3]]
+// CHECK-DAG:  %[[C_01_1:.+]] = amdgpu.mfma %[[A_EXTRACT00]] * %[[B_EXTRACT10]] + %[[C_01_0]]
+// CHECK-DAG:  %[[C_01_2:.+]] = amdgpu.mfma %[[A_EXTRACT01]] * %[[B_EXTRACT11]] + %[[C_01_1]]
+// CHECK-DAG:  %[[C_01_3:.+]] = amdgpu.mfma %[[A_EXTRACT02]] * %[[B_EXTRACT12]] + %[[C_01_2]]
+// CHECK-DAG:  %[[C_01_4:.+]] = amdgpu.mfma %[[A_EXTRACT03]] * %[[B_EXTRACT13]] + %[[C_01_3]]
+// CHECK-DAG:  %[[C_70_1:.+]] = amdgpu.mfma %[[A_EXTRACT70]] * %[[B_EXTRACT00]] + %[[C_70_0]]
+// CHECK-DAG:  %[[C_70_2:.+]] = amdgpu.mfma %[[A_EXTRACT71]] * %[[B_EXTRACT01]] + %[[C_70_1]]
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_tile_and_fuse.mlir:815]: ditto, can add `CHECK-COUNT` to make sure there are the appropriate number of mfma ops.

**Review by Max191 — COMMENTED** [review]
This is very clean with the enums! Please add some docs to the helper functions.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp (hunk 4 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: We are already asserting in `getIntrinsicSwizzle`, at the root cause.  I will just add TODO comments.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_tile_and_fuse.mlir (hunk 5 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_tile_and_fuse.mlir:795]: The problem is that these different kinds of lines are interleaved, which is why CHECK-DAG helps here.

**Comment by bjacob** [comment]
I had to move things around to break a Bazel circular dependency. That's the last commit here.

**Review by hanhanW — COMMENTED** [review]
cool stuff!

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 7]
```diff
@@ -903,10 +903,33 @@ std::tuple<int64_t, int64_t, int64_t> DataTiledMMAAttr::getMNKShape() const {
           opaqueLayout.kSize * getUnrollK()};
 }
 
+// Returns the swizzled tile shape, but with dim sizes overwritten with 1 if
+// `predicate` returns false.
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: Use `///` outside of functions

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 8]
```diff
@@ -917,6 +940,169 @@ FailureOr<IREE::GPU::MMAScope> DataTiledMMAAttr::getMmaScope() const {
...
+
+  offsets.append(tileOffsets);
+  sizes.append(tileSizes);
+  strides.append(tileStrides);
+
+  return success();
+}
+
+// Increment the mutable vector `indices` to traverse the index space below
+// `sizes`, with the last dimension moving fastest.
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: use `///`

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 9]
```diff
@@ -917,6 +940,169 @@ FailureOr<IREE::GPU::MMAScope> DataTiledMMAAttr::getMmaScope() const {
...
+    ++indices[i];
+    if (indices[i] == sizes[i]) {
+      indices[i] = 0;
+    } else {
+      break;
+    }
+  }
+}
+
+// Flattens the input vector `value` to 1-D.
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: use `///`

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 10]
```diff
@@ -917,6 +940,169 @@ FailureOr<IREE::GPU::MMAScope> DataTiledMMAAttr::getMmaScope() const {
...
+  if (vectorType.getRank() <= 1) {
+    return value;
+  }
+  VectorType flatVectorType =
+      VectorType::get(SmallVector<int64_t>{vectorType.getNumElements()},
+                      vectorType.getElementType());
+  return builder.create<vector::ShapeCastOp>(loc, flatVectorType, value);
+}
+
+// Returns intrinsic-level slices tiling the input multi-MMA-level tile `value`.
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: use `///`

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 11]
```diff
@@ -917,6 +940,169 @@ FailureOr<IREE::GPU::MMAScope> DataTiledMMAAttr::getMmaScope() const {
...
+    Value extract = builder.create<vector::ExtractStridedSliceOp>(
+        loc, value, indices, internalShape, strides);
+    distributedValue = flattenVector(builder, loc, extract);
+    incrementIndices(indices, crossIntrinsicShape);
+  }
+  return distributedValues;
+}
+
+// Generates amdgpu.mfma/wmma operation on the given inputs for this attribute
+// type.
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: use `///`

**Review by Max191 — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 12]
```diff
@@ -903,10 +903,33 @@ std::tuple<int64_t, int64_t, int64_t> DataTiledMMAAttr::getMNKShape() const {
...
 }
 
+// Returns the swizzled tile shape, but with dim sizes overwritten with 1 if
+// `predicate` returns false.
+static SmallVector<int64_t>
+sliceSwizzledShape(const TileSwizzle &swizzle,
+                   std::function<bool(TileSwizzle::Dim)> predicate) {
+  SmallVector<int64_t> shape;
+  for (auto e : swizzle.expandShape) {
+    for (auto d : e) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: nit: Use actual types here -- it's hard to know what these are based on the names only

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp [hunk 13]
```diff
@@ -917,6 +940,169 @@ FailureOr<IREE::GPU::MMAScope> DataTiledMMAAttr::getMmaScope() const {
...
+            builder, loc, intrinsicCType, intrinsicsLhs[mu * getUnrollK() + ku],
+            intrinsicsRhs[nu * getUnrollK() + ku],
+            intrinsicsAcc[mu * getUnrollN() + nu]);
+      }
+    }
+  }
+
+  // Insert the results into the destination accumulator.
+  auto accSwizzle = getSwizzle(*this, MMAFragment::Acc);
+  auto accCrossIntrinsicShape =
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.cpp]: also here: not sure what the type is

**Pull request merged**